crates/curp/src/tracker.rs

```diff
@@ -0,0 +1,351 @@
+#![allow(unused)] // TODO remove when used
+
+use std::collections::VecDeque;
+use std::ops::{AddAssign, Sub};
+
+use clippy_utilities::NumericCast;
+
+/// Bits of usize
+const USIZE_BITS: usize = std::mem::size_of::<usize>() * 8;
+
+/// Default bit vec queue capacity, this is the number of inflight requests that a client expects.
+/// It can reduce about `log_2(DEFAULT_BIT_VEC_QUEUE_CAP)` `VecDequeue` memory reallocations.
+const DEFAULT_BIT_VEC_QUEUE_CAP: usize = 1024;
+
+/// A one-direction bit vector queue
+/// It use a ring buffer `VecDeque<usize>` to store bits
+///
+/// Memory Layout:
+///
+/// `010000000101_110111011111_000001000000000`
+///  ^        ^                ^    ^
+///  |        |________________|____|
+///  |________|         len    |____|
+///     head                    tail
+#[derive(Debug, Clone)]
+struct BitVecQueue {
+    /// Bits store
+    store: VecDeque<usize>,
+    /// Head length indicator
+    head: usize,
+    /// Tail length indicator
+    tail: usize,
+}
+
+impl Default for BitVecQueue {
+    #[inline]
+    fn default() -> Self {
+        Self::with_capacity(DEFAULT_BIT_VEC_QUEUE_CAP)
+    }
+}
+
+#[allow(clippy::arithmetic_side_effects, clippy::unwrap_used)] // They are checked
+impl BitVecQueue {
+    /// New with a capacity (in bits)
+    fn with_capacity(cap_bits: usize) -> Self {
+        Self {
+            store: VecDeque::with_capacity(cap_bits / USIZE_BITS),
+            head: 0,
+            tail: 0,
+        }
+    }
+
+    /// Get the length
+    fn len(&self) -> usize {
+        if self.store.is_empty() {
+            return 0;
+        }
+        (self.store.len() - 1) * USIZE_BITS + self.tail - self.head + 1
```

Comment:
Is `self.tail` always greater than or equal to `self.head + 1`?

Reply:
No, head could equals to tail, when they are both 0, which means that the bitvec is empty.